### DISCUSSION TITLE
Running into issues with content security policy

### DISCUSSION THREAD

**Discussion by tmoleary21** [uber/react-map-gl#2027] [Q&A] [answered]
Hello,

My team is working on migrating our data visualization tool built in React (w/ webpack) from Leaflet to Deck.gl to take advantage of the graphics acceleration. We are using react-map-gl to provide a base map. We are a research team at a university, which means we have a strict content security policy we have to adhere to, so as I've been trying to deploy our app I've run into issues with mapbox trying to create workers from a blob.

We have been using react-map-gl v5 with mapbox-gl to avoid the auth token, but since working on this issue I switched to the latest release with maplibre. 

Below is the file that does everything Deck.gl and react-map-gl related. Far from our whole app, but nothing else deals with these packages.

```
import React, {useState} from 'react';
import DeckGL from '@deck.gl/react';
import { BASEMAP } from '@deck.gl/carto';
import MapFromRMG from 'react-map-gl';
import maplibreGl from 'maplibre-gl';
import { WebMercatorViewport } from '@deck.gl/core';
import GoTo from '../widgets/GoTo';
import LayerManager from "./LayerManager";
import {debounce} from "@mui/material";

export default function Map(props){

    const [initialViewState, setInitialViewState] = useState({
        longitude: -105.08,
        latitude: 40.56,
        zoom: 10,
        pitch: 0,
        bearing: 0
    });

    const [layers, setLayers] = useState([]);

    // Pass the setter to LayerManager so it can automatically update the rendered layers
    LayerManager.setLayersState = setLayers;

    // eslint-disable-next-line import/no-webpack-loader-syntax
    const workerClass = require('worker-loader!mapbox-gl/dist/mapbox-gl-csp-worker').default;

    return (
        <>
            <DeckGL
                initialViewState={initialViewState}
                controller={{dragRotate: false}}
                onViewStateChange={ debounce(({viewState}) => {
                    const viewport = new WebMercatorViewport(viewState);
                    //Note that the unproject function does not bound to [-180,180] degrees
                    const nw = viewport.unproject([0, 0]);
                    const se = viewport.unproject([viewport.width, viewport.height]);

                    LayerManager.updateBounds({
                        _northEast:{lat:nw[1],lng:se[0]},
                        _southWest:{lat:se[1],lng:nw[0]}
                    }, viewState.zoom);
                }, 100)}
                layers={layers}
            >
                <MapFromRMG mapStyle={BASEMAP.VOYAGER} mapLib={maplibreGl} workerClass={workerClass}/>
            </DeckGL>
            <div id="current-location" className="current-location">
                <GoTo setInitialViewState={setInitialViewState}/>
            </div>
        </>
    );
}
```

Errors:
<img alt="image" src="https://user-images.githubusercontent.com/43046771/198516872-094008b9-c87d-4183-88d0-f23a5f464076.png">

The latest thing I have tried is setting the `workerClass`. The docs say that setting the `workerUrl` is useful for handling CSP issues, but I couldn't figure out what the URL should be. I thought the `workerClass` would do the trick, but that could be wrong or I could be setting it incorrectly. To reiterate, I only get this issue in production (CSP only matters in production for us).

Any help with fixing this would be greatly appreciated!

A subset of the resources that have helped:
https://docs.mapbox.com/mapbox-gl-js/guides/browsers-and-testing/#strict-csp-environments
https://visgl.github.io/react-map-gl/docs/api-reference/map

**Comment by Pessimistress** [comment]
If you want to use maplibre, then you need to look at maplibre's documentation.

https://maplibre.org/maplibre-gl-js-docs/api/#csp-directives

**Reply by tmoleary21** [reply]
Swapping the "mapbox-gl" to "maplibre-gl" in the require statement didn't change anything, except that I am using the correct packages now (which is good). 

What I've tried since this post:
* Using an import statement instead of require
* Getting the class without the worker-loader both with require and import
* Hosting the app directly from my college's domain to ensure the blob is the issue, not the url

Could any of this be fixed by using `workerUrl` instead?

Do you know of any examples of projects using react-map-gl in a strict CSP environment? I've seen related issues and somewhat similar stackoverflow posts, but I haven't found any good examples to refer to.

**Reply by Pessimistress** [reply]
I'm not sure what you are trying to achieve by setting `workerClass`. The documentation is pretty clear that you need to set `workerUrl`.

**Reply by tmoleary21** [reply]
Understood, I'll give `workerUrl` another try. I'll make another reply if I make any actual progress

**Reply by tmoleary21** [reply]
Here's how I'm using workerUrl now:
```
<MapFromRMG mapStyle={BASEMAP.VOYAGER} mapLib={maplibreGl} workerUrl={new URL("/node_modules/maplibre-gl/dist/maplibre-gl-csp-worker.js", import.meta.url)}/>
```
I'm not sure that it makes sense to use the path through node_modules to get to the csp worker file, but that is my best guess based on the documentation [here](https://visgl.github.io/react-map-gl/docs/api-reference/map) and [here](https://maplibre.org/maplibre-gl-js-docs/api/#csp-directives). I can't use the unpkg URL of course because of CSP. 

Do you know what I'm doing wrong? Do I need to somehow use the `maplibre-gl-csp.js` file as well, like it does in the docs?

**Reply by Pessimistress** [reply]
Ok I finally understand why you are so confused. The unpkg url from the documentation just tells you which file you need to feed to the `workerUrl` prop. In practice you need to copy that file out to your own application and serve it as a static asset, so that it can be loaded through an accessible URL such as `https://myuniversity.edu/app/static/maplibre-gl-csp-worker.js`.

**Reply by tmoleary21** [reply]
Alright, thank you that definitely helps. It's still not working but I'm starting to see some issues on my end that I will need to address before I decide to come back here. I appreciate your help!

**Reply by tmoleary21** [reply]
Hi again! I'm still having the same issues on Google Chrome. I am providing a `workerUrl` to the Map component that points to the `maplibre-gl-csp-worker.js` file on our server. It seems like it should be working properly because it works in Firefox, and I can see it is making network requests for the workers using the URL I provided. On Chrome, I still get the same error as in my first post, and there is no evidence it is using the URL at all.

I've also tried changing my MapLibre import to use the strict csp version. The documentation just says that changing workerUrl should do the trick, but I thought I would try it: `import MaplibreGl from 'maplibre-gl/dist/maplibre-gl-csp';`

Any new advice would be appreciated.

**Reply by tmoleary21** [reply]
I am back to working on this again after a few months working on other things. Still having the same issues described above.

I have two versions running on our server now. One that is running only maplibre-gl without npm, react, or anything extra. It is similar to the maplibre [Display a map](https://maplibre.org/maplibre-gl-js-docs/example/simple-map/) example, but with the [strict csp environment](https://maplibre.org/maplibre-gl-js-docs/api/#csp-directives) fix. The other version is our normal codebase with react-map-gl. For that one, I am passing the maplibre object imported from maplibre-gl-csp.js to the `mapLib` prop. In both cases, I am passing the same URL to the static asset maplibre-gl-csp-worker.js to the `workerUrl`. The vanilla JavaScript version is working, the react app is not.

Could this be an issue with react-map-gl? Do you think I am still doing something wrong, or could it be because of some difference from using npm/react/webpack/etc? I appreciate the help.

**Reply by tmoleary21** [reply]
I finally found the issue around a week ago. In map.tsx, `mapboxgl.supported` is called before the package is used for anything else, but the supported function creates a worker from a blob URL within it to check for worker support, which violates the content security policy. **I fixed it with `maplibregl.supported = () => true;`** I see this as a temporary solution. Hopefully the contributors of [mapbox-gl-supported](https://github.com/mapbox/mapbox-gl-supported) can add a fix for the supported function that accounts for this

**Comment by tmoleary21** [comment, marked as answer]
The `mapboxgl.supported` function creates a worker from a blob URL to check worker support, regardless of whether you are using the CSP compliant alternatives. `mapboxgl.supported = () => true` will remove the violations.

**Reply by Pessimistress** [reply]
Thanks for the insight. #2138 should fix this.
